documentation/Calls.md

```diff
@@ -0,0 +1,443 @@
+information used by the runtime support library for initialization,
+copying, finalization, and I/O of type instances).  Each overridable
+type-bound procedure in the type corresponds to an index into this table.
+
+### Host association linkage
+
+Calls to dummy procedures and procedure pointers that resolve to
+internal procedures need to pass an additional argument that
+addresses on block of storage in the stack frame of the their
+host subprogram that was active at the time they were passed as an
+actual argument or associated with a procedure pointer.
+This is similar to a static link in implementations of programming
+languages with nested subprograms, although Fortran only allows
+one level of nesting.
+The 64-bit x86 and little-endian OpenPower ABIs reserve registers
+for this purpose (`%r10` & `R11`); 64-bit ARM has a reserved register
+that can be used (`x18`).
+
+The host subprogram objects that are visible to any of their internal
+subprograms need to be resident in memory across any calls to them
+(direct or not).  Any host subprogram object that might be defined
+during a call to an internal subprogram needs to be reloaded after
+a call or reside permanently in memory.
+A simple conservative analysis of the internal subprograms can
+identify all of these escaping objects and their definable subset.
+
+The address of the host subprogram storage used to hold the escaping
+objects needs to be saved alongside the code address(es) that
+represent a procedure pointer.
+It also needs to be conveyed alongside the text address for a
+dummy procedure.
+
+For procedures that can be called with an implicit interface,
+we cannot use a "procedure pointer descriptor" to pass them an
+a procedure argument -- a Fortran '77 routine
+with an `EXTERNAL` dummy argument expects to receive a single
+address argument.  Instead, when passing an actual procedure
+to a procedure that can be called with an implicit interface,
+we will need to package the host link in a trampoline.
+
+GNU Fortran and Intel Fortran construct trampolines by writing
+a sequence of machine instructions to a block of storage in the
+host's stack frame, which requires the stack to be executable,
+which seems inadvisable for security reasons;
+XLF manages trampolines in its runtime support library, which adds some overhead
+to their construction and a reclamation obligation;
+NAG Fortran manages a static fixed-sized stack of trampolines
+per call site, imposing a hidden limit on recursion and foregoing
+reentrancy;
+PGI passes host variable links in descriptors in additional arguments
+that are not always successfully forwarded across implicit interfaces,
+sometimes leading to crashes when they turn out to be needed.
+
+F18 will use descriptors in most cases to pass procedures as arguments,
+but also maintain a pool of trampolines in its runtime support library
+to implement calls to routines that need to support an
+implicit interface.
+
```

Comment:
Will this design allow for cases like the one mentioned in the following ticket? Here the C program calls a Fortran subroutine 'c', passing it a pointer to the subroutine 'b'. Subroutine 'c' calls 'b'. (Not supported by Flang but works with gfortran and Intel. We are told it is a common pattern) 
https://github.com/flang-compiler/flang/issues/631